### ISSUE TITLE
broken on torsocks

### ISSUE THREAD

**Issue by bruceleerabbit** [Genymobile/gnirehtet#558]
This works fine, which is the normal case over clearnet:
```
GNIREHTET_APK=/my_apk_files/gnirehtet_v25.apk /usr/local/bin/gnirehtet run
```
But if `gnirehtet` is run on torsocks to proxy over tor, it falls apart because gnirehtet needs to make local connections:
```
$ GNIREHTET_APK=/my_apk_files/gnirehtet_v25.apk torsocks /usr/local/bin/gnirehtet run
... WARNING torsocks[…]: [connect] Connection to a local address are denied since it might be a TCP DNS query to a local DNS server. Rejecting it for safety reasons. (in tsocks_connect() at connect.c:191)
error: could not install *smartsocket* listener: Address already in use

* failed to start daemon
error: cannot connect to daemon
ERROR Main: Cannot stop client: Command adb ["shell", "am", "start", "-a", "com.genymobile.gnirehtet.STOP", "-n", "com.genymobile.gnirehtet/.GnirehtetActivity"] returned with value 1
```
This is related to bug #329 by @Kardshadow.  The readme shows some manual ways of running gnirehtet. I’m sure just one of those commands needs to run on torsocks, but whichever command that is, it’s likely also broken in this regard because it will need to talk locally.

Gnirehtet really needs specific support for proxying.

It’s an important feature. Mind boggling that it was not built into the original design because Android devices are “chatty”. When running gnirehtet it’s immediately clear that there are unnecessary connections phoning home to Google. So the very first concern should be a way to anonymise that traffic and also to (egress) firewall it to stop the data leaks.

**Comment by VaradGupta23** [comment]
You're right — running gnirehtet through torsocks breaks it, because gnirehtet is designed to:

Communicate with the Android device over local ADB sockets

Listen on a local port (31416) for VPN-tunneled traffic from the device

torsocks, by design, blocks all local (127.0.0.1) traffic — to prevent accidental identity leaks through direct local DNS or loopback connections — which is exactly what gnirehtet depends on.

Why It Fails
The key line:

torsocks: Connection to a local address are denied… Rejecting it for safety reasons.

So when you do:


torsocks gnirehtet run
…it fails because:

adb uses local sockets

gnirehtet relay listens on localhost:31416

torsocks blocks all of this for anonymity safety

 Workaround (Partial): Only Proxy Traffic, Not gnirehtet
 Run gnirehtet normally (without torsocks), but force device traffic to go through Tor using:
1. Tor as transparent proxy (recommended)
Set up iptables on the host (your Linux PC) to NAT and redirect all outbound traffic (from Android via gnirehtet) through the Tor daemon listening on port 9040.

2. Use redsocks or proxychains
Wrap traffic from the gnirehtet interface (typically a tun interface) through a local socks5 proxy like Tor using redsocks.

🛠️ Conceptual Setup (Tor Routed gnirehtet Traffic)
Run gnirehtet normally (no torsocks):


GNIREHTET_APK=./gnirehtet.apk ./gnirehtet run
Redirect traffic from 10.0.0.2 (device via gnirehtet VPN) → Tor SOCKS

You can use:

iptables + redsocks

Or use pf/nftables depending on your OS

Example (Linux):

sudo iptables -t nat -A OUTPUT -p tcp -s 10.0.0.2 -j REDIRECT --to-ports 12345
Where port 12345 is your redsocks local port configured to use Tor’s SOCKS proxy.

 What You Cannot Do
You cannot run gnirehtet itself under torsocks

You cannot use ADB via torsocks (adb is 100% local-socket-based)

Alternatives
If full anonymity is your goal, consider:

Using Orbot VPN directly on Android (no root needed)

Or using GrapheneOS or CalyxOS with built-in DNS + traffic hardening

Using iptables or nftables to isolate Android traffic manually

Security Notes
You're absolutely right:

Gnirehtet traffic is unencrypted

Android can be “chatty” with Google, even over USB reverse tethering

Best practice is to inspect traffic with tcpdump or Wireshark and use transparent proxying tools like redsocks to anonymize device egress.
